### ISSUE TITLE
Inference issue using generics with sealed class

### ISSUE THREAD

**Issue by gabrielgarciagava** [dart-lang/sdk#52748]
The inferrer fails to infer the type of the generics of a function

I created a small sample here https://dartpad.dev/?id=61a4d13af42ac7e6749333728587fa08

Issue reproduced using dart 3.0.5

**Comment by gabrielgarciagava** [comment]
Hmm, I just realised that it I can "workaround" it by specifying the foo function as:
```
bool foo<G extends Generics<B>, V extends B, B>(G g, V v) {
  return g.value == v;
}
```
instead of
```
bool foo<G extends Generics<V>, V>(G g, V v) {
  return g.value == v;
}
```
Should I assume this is actually the expected behaviour?

**Comment by srawlins** [comment]
Yes I believe this is working as intended. In which case you can open an issue at https://github.com/dart-lang/language/issues, if you think the algorithm could be improved. For extra points, you could search that issue tracker for a similar inference feature request, as there are many inference feature requests :)

**Comment by eernstg** [comment]
I think it could be useful to give a comment on the situation (considered as an SDK issue, not diving into language design), so I'll add a comment here even though the issue is closed.

First, pragmatically, I'd recommend using a declaration of `foo` that does not have the type variable `G`:

```dart
bool foo<V>(Generics<V> g, V v) => g.value == v;
```

The point is that the introduction of an extra type variable `G` (in the original version of `foo`) doesn't yield any improved typings anywhere: In the body of `foo`, the interface of a receiver of type `G` is the same as the interface of its bound `Generics<V>` anyway, and `G` isn't used in the return type. So neither the body of `foo` nor the callers of `foo` will get a better typing by having that extra type variable. Hence, the obvious rule of thumb is to avoid such unnecessary type variables.

Next, we could ask whether the type inference outcome is as expected. The reason why type inference fails is, essentially, that downwards inference on the invocation of `foo` does not carry any information, but the actual arguments (of type `Generics<Base>` and `Impl`) bind `G` to `Generics<Base>` and `V` to `Impl`, and it is then not true that `G <: Generics<V>`.

You can see in more detail how constraints are generated and solved [here](https://github.com/dart-lang/language/blob/main/resources/type-system/inference.md#grounded-constraint-solution-for-a-set-of-type-variables). The main thing to note is that there is no way for information to flow from a type variable like `V` into a type variable like `G` based on the fact that `V` occurs in the bound for `G`. So `V` will be chosen based on other things, and then we'll _check_ that `G` is a subtype of the bound which depends on `V`.
